Initial week schedule

MattWindsor91 opened the pull request:
This PRQ pulls in the week schedule implementation I've been working on, on and off (mostly off), for the past year (!).

This is _not_ the final version of the week schedule.  Missing features/areas for improvement include:

- Appropriate style (delegated to the frontend team)
- Schedule block colouring
- Routes for today's schedule and other 'relative' schedules
- Day schedules (if we're still doing these)
- The implementation could do with cleanup and optimisation, eg. doing hour culling calculations on the first and last show of each day rather than through the entire schedule.  See @DinCahill's week schedule implementation, which does a lot more day breaking-up than mine does.
- Unit tests

This PRQ mainly exists to try and get a schedule into 2016-site ASAP (after code review, of course!) so it can start being styled and integrated, while we work out how to improve the actual codebase, probably including ideas from @DinCahill's implementation.

MattWindsor91 committed:
Add initial boilerplate for week schedule.

This won't work yet...

MattWindsor91 committed:
Plumb in (useless) controller.

MattWindsor91 committed:
Begin week schedule template.

MattWindsor91 committed:
Get a working weekly schedule call.

MattWindsor91 committed:
Add some schedule types in preparation for filling.

This is almost certainly going to need to change around a lot,
the way this drills into the config is a bit nasty.

MattWindsor91 committed:
Finish adding filling.

MattWindsor91 committed:
Implement schedule formatter.

This is probably going to be replaced with Din's formatter, but
I thought I'd complete it anyway just in case.

It seems to handle DST properly, but won't handle schedule bugs
gracefully.

MattWindsor91 committed:
Remove magic number.

MattWindsor91 committed:
Fix bugs in schedule boundary detection.

Also refactor a few bits of the algorithm.

MattWindsor91 committed:
Use %d, not %i, for integer formats in errors

This was a typo, and caused error messages to be less human
readable than they should be.

MattWindsor91 committed:
Use startOffsetToHour in week schedule calculator

For consistency and ease of changing later.

MattWindsor91 committed:
Move and tidy up date/time logic

Generic date manipulation functions now live in utils/date.go.
I've also run gofmt on some of the existing code in the process.

MattWindsor91 committed:
Make start hour private

This will help if we ever decide to make the start hour
configurable.

MattWindsor91 committed:
Refactor and fix empty schedule bug

This algorithm now appropriately sets the schedule table to nil
when there are no non-sustainer shows.

MattWindsor91 committed:
Keep entire day range in WeekSchedule

WeekSchedule now tracks the entire set of days it manages, as
opposed to just the start and end day.  It also uses URY start
times in more places, as opposed to local midnight.

This simplifies a bit of the logic, as well as making us able to
reduce the amount of hardcoding in the template.

Technically, this means we could generate 9-day schedules (or
n-day schedules for arbitrary n), but whether this is a good thing
is debatable.

MattWindsor91 committed:
Fix minor template inconsistency

MattWindsor91 committed:
Export and use show descriptions

This isn't quite right---show descriptions are arbitrary HTML, and
we currently just escape it, which is ugly.

MattWindsor91 committed:
Strip HTML in tooltip descriptions

MyRadio sends show descriptions as full-on HTML5, which is great
for show description pages, but bad for tooltips.  This commit
teaches 2016-site to strip the HTML when emitting descriptions in
tooltips.

This is a fairly heavyweight way to do this, involving manually
invoking a HTML parser, but it seems to work ok.

MattWindsor91 committed:
Add timeslot links to week schedule

These use the new url function and thus respect prefixes.

MattWindsor91 committed:
gofmt: fix minor space error

MattWindsor91 committed:
Use router reversing for timeslot URLs

This means we don't need to construct the URLs ourselves, which
saves repetition and a source of errors later on, but _does_ mean
we have to pass the router around into places from which it used to
be separate.

This is a bit more invasive than I'd like, and could probably be
refactored a lot.

MattWindsor91 committed:
goimports: Rearrange imports

MattWindsor91 commented on config.toml.example:
```diff
@@ -29,6 +29,10 @@
 	newsTwitter = "URYNews"
 
+	[pageContext.sustainer]
+		name = "Overnight Owen"
```
IDK if people want me to change this default to `URY Jukebox` or something.  I mainly put it here as an outdated joke.

LordAro replied:
Currently the deploy script just copies the example file, so it'd certainly make it easier to deploy, but making it different would force us to actually do something properly...

MattWindsor91 commented on controllers/schedule_week.go:
```diff
@@ -0,0 +1,400 @@
+package controllers
```
Should this be part of the show controller?

LordAro replied:
I don't see why it should be

MattWindsor91 commented:
I'll fix the codacy nits soonish (but probably not tonight)

MattWindsor91 commented:
Also son of a submariner is codacy infuriatingly condescending

MattWindsor91 committed:
Fix Codacy linter errors

There might be more linter error on the way, but these were the
ones reported by
https://www.codacy.com/app/hatton.brooke/2016-site/pullRequest?prid=732694

MattWindsor91 committed:
Refactor the week schedule tabulator

Split some functions up, especially in terms of separating the
schedule generation from the template rendering.

MattWindsor91 commented on controllers/schedule_week.go:
```diff
@@ -0,0 +1,409 @@
+func weekdayFromVars(vars map[string]string) (string, string, string, error) {
+	y, ok := vars["year"]
+	if !ok {
+		return "", "", "", errors.New("no year provided")
```
This might be a good place to use named return values.  I'm torn.

LordAro replied:
I'm tempted to say yes here

MattWindsor91 commented on controllers/schedule_week.go:
```diff
@@ -0,0 +1,409 @@
+
+//
+// Week schedule algorithm
+// TODO(CaptainHayashi): move?
```
It's difficult to tell where most of this should be.  There's also some code in `structs/schedule.go` that probably belongs in the same place as this

MattWindsor91 commented on controllers/schedule_week.go:
```diff
@@ -0,0 +1,409 @@
+		return
+	}
+
+	// These are the boundaries for culling, and are expanded upwards when we find shows that start earlier or finish later than the last-set boundary.
```
This entire bit of code could be made a lot more efficient if we hold onto the first and last shows of each day.

LordAro replied:
How much more efficient? Why don't we hold onto these shows?

MattWindsor91 commented on controllers/schedule_week.go:
```diff
@@ -0,0 +1,409 @@
+}
+
+// tabulateWeekSchedule creates a schedule table from the given schedule slice.
+func tabulateWeekSchedule(start, finish time.Time, schedule []structs.ScheduleItem) (*WeekSchedule, error) {
```
These function names might be a bit overlong

MattWindsor91 commented on models/schedule_week.go:
```diff
@@ -0,0 +1,30 @@
+	"github.com/UniversityRadioYork/myradio-go"
+)
+
+// ScheduleWeekModel is the model for the ScheduleWeek controller.
```
This model seems too specific.  The other thing I'm unsure of is whether any of the business logic from above should be in here, but I doubt it.

MattWindsor91 commented on models/schedule_week.go:
```diff
@@ -0,0 +1,30 @@
+// On success, it returns the day-split map of timeslots in the week schedule, and nil.
+// Otherwise, it returns undefined data and the error causing failure.
+func (m *ScheduleWeekModel) Get(year, week int) (map[int][]myradio.Timeslot, error) {
+	// TODO(CaptainHayashi): Jukebox filling
```
This is handled elsewhere, so I should probably bin this comment

MattWindsor91 commented on structs/config.go:
```diff
@@ -30,10 +30,14 @@ type PageContext struct {
 	MainFacebook    string `toml:"mainFacebook"`
 	NewsTwitter     string `toml:"newsTwitter"`
 	Pages           []Page
+	Sustainer       struct {
```
Not sure if `Sustainer` is the best term to use, but it's the most brand-neutral.

MattWindsor91 commented on structs/schedule.go:
```diff
@@ -0,0 +1,248 @@
+	HasPage() bool
+
+	// GetPageURL gets the root-relative URL to this schedule item's page.
+	GetPageURL() string
```
Maybe this should return `(string, bool)` instead of `HasPage` being a separate thing.  Or even `(URL, bool)`

MattWindsor91 commented on structs/schedule.go:
```diff
@@ -0,0 +1,248 @@
+
+			firstFinish := first.StartTime.Add(first.Duration)
+			if firstFinish.After(second.StartTime) {
+				return nil, fmt.Errorf(
```
We should probably recover more gracefully from this, as it happens a lot in real life

MattWindsor91 commented on utils/date.go:
```diff
@@ -0,0 +1,123 @@
+//
+
+// The hour of the day (local time) at which the scheduled day begins.
+const startHour = 6
```
This could be a configurable later on.

MattWindsor91 commented on utils/date.go:
```diff
@@ -0,0 +1,123 @@
+func ParseIsoWeek(year, week, weekday string) (int, int, time.Weekday, error) {
+	y, err := strconv.Atoi(year)
+	if err != nil {
+		return 0, 0, 0, err
```
Named return values?

LordAro replied:
I think named return values would be good, if not just to explain what the 4 return values are

MattWindsor91 commented on utils/date.go:
```diff
@@ -0,0 +1,123 @@
+	// ISOWeek on it.
+	fj := time.Date(year, time.January, 4, 0, 0, 0, 0, time.Local)
+
+	// Correct Go's stupid Sunday is 0 decision, making the weekdays ISO 8601 compliant
```
Welp, I didn't mince words a year ago

MattWindsor91 commented on utils/html.go:
```diff
@@ -0,0 +1,68 @@
+const stripAllowedNodes = "html body p a strong em span section div"
+
+// StripHTML strips HTML tags from a string, extracting all plain text.
+func StripHTML(htmls string) (string, error) {
```
This is horrendous and mostly exists to stop the horrors of the HTML-based show description from screwing up tooltips.

Brookke replied:
Can we not just strip html from show descriptions anyway and make them all plain text, in myRadio? That method would be my preferred fix

MattWindsor91 replied:
I would definitely be for this.

Brookke replied:
@MattWindsor91 lets make it happen :smile:

MattWindsor91 commented:
Recommend comparing and working in @DinCahill's [week schedule](https://github.com/DinCahill/2016-site/tree/din-schedule).

MattWindsor91 commented:
One thing @DinCahill does that I haven't is put all of the schedule padding logic in `myradio-go`, instead of here.  Thoughts?

My gut feeling is that if anything else (eg a future rewrite of myradio) needs schedule logic from here, then we can split it out into a separate library.  But if people think it'd be better in `myradio-go`, either now or at that point, then sure.

Brookke commented:
I would leave it where it is personally @MattWindsor91. Could we look into generating myradio-go from swagger I believe @LloydW93 has some knowledge of this

MattWindsor91 committed:
Add 'this week' schedule endpoint

Link to the endpoint from navbar.

Lightly refactor the controller to make this possible without duplication.

Also, make some variable names cleaner.

MattWindsor91 committed:
Add previous/next links to schedule pages

Routing these needs care, as the week schedule routes aren't available when the URL reversers are installed.

MattWindsor91 committed:
Prettify week schedule placeholder template

I haven't seen the design for the week schedule yet, so this is still just a placeholder.

It vaguely looks presentable now, though.

MattWindsor91 committed:
Add fancy week name formatting

This is a straight port of the logic from 2013-site, but needs some
unit testing over here.

MattWindsor91 committed:
Reduce ScheduleItem to one struct

There's no need for an interface and two structs when the only difference is data.

Also move the sustainer context passing from template to controller level.

This removes/generalises some template funcs.

MattWindsor91 committed:
Add placeholder styling for schedule blocks

This currently only picks out sustainer and non-sustainer, and
does so in a rather ugly way, but you get the gist.

MattWindsor91 committed:
Move sustainer config to [schedule] key

This cleans up the page context and gives us a logical place to put further schedule config later.

MattWindsor91 committed:
sass: Make codacy happy

Probably make someone else happy, too.

Some day.

MattWindsor91 commented:
tfw making codacy happy doesn't make codacy happy

MattWindsor91 committed:
More appeasing of the Codacy winged monkeys

FLY!  FLY!  FLY!!

MattWindsor91 committed:
Move anon row-decision struct to top level

This means we can split out the row deciding code a bit more later.

MattWindsor91 committed:
Aggressively refactor schedule_week

Make difference between start offsets and hours obvious with newtypes.

Split row generator code up into multiple functions.

MattWindsor91 committed:
Clean up showStraddlesDay

I'd move this to structs as a method on ScheduleItem, but that
would create a cyclic dependency between utils and structs.

MattWindsor91 committed:
Fix off-by-one bug on bottom offset calculation

The bottom offset was being set to the hour the last show finishes,
which is not the same as the last row the show occupies when the
show finishes exactly on the hour.

This could do with a unit test.

MattWindsor91 commented:
Interesting URY schedules for bug testing (I'll add to these later):

Year | Week | Reason | Bugged? | Sitelinks
---- | ---- | ------ | ------- | ---------
1 | 1 | Crashes 2013-site; not allowed in 2016-site | No? | [C(S)](https://ury.org.uk/schedule/1/w1/)
1000 | 1 | Earliest schedule permitted by 2016-site routing.  Crashes due to failed URL reverse. | [Fixed](https://github.com/UniversityRadioYork/2016-site/pull/64/commits/a2a563153e07e90cff92a53cc0e20cf2b194ebef) | [C](https://ury.org.uk/schedule/1000/w1/)
2004 | 17 | Crashes 2013-site | No | [C(S)](https://ury.org.uk/schedule/2004/w17/)
2004 | 18 | Immediately after above; crashes due to an overlap between timeslot 'Overnight Music', ID 118482, finishing at 2004-05-02 09:00:00 +0100 BST and 'Overnight Music', ID 118476, starting at 2004-04-26 02:00:00 +0100 BST' | [Fixed](https://github.com/UniversityRadioYork/2016-site/pull/64/commits/a34c5c7d7384aef41730ef1b2cbb492d5f010556) | [C](https://ury.org.uk/schedule/2004/w18/)
2007 | 04 | Show starting at half-past: Backslash (NB: the problem with The Heidi & Jo Show is _not_ a bug) | No | [C](https://ury.org.uk/schedule/2007/w04/)
2008 | 18 | Shows end at :59; crashes 2013-site | No | [C(S)](https://ury.org.uk/schedule/2008/w18/)
2008 | 25 | Shows end at :59; crashes 2013-site | No | [C(S)](https://ury.org.uk/schedule/2008/w25/)
2010 | 09 | Last show on a night ends on a half-hour: note 2013-site culls the half-hour row, which is arguably incorrect | No | [C(B)](https://ury.org.uk/schedule/2010/w09/)
2010 | 22 | Crashes 2013-site; crashes due to an overlap between timeslot 'The Lukebox', ID 122059, finishing at 2010-06-01 16:00:00 +0100 BST and 'Campus Chats with Chris Young', ID 114448, starting at 2010-06-01 15:00:00 +0100 BST' | [Fixed](https://github.com/UniversityRadioYork/2016-site/pull/64/commits/a34c5c7d7384aef41730ef1b2cbb492d5f010556)| [C(S)](https://ury.org.uk/schedule/2010/w22/)
2010 | 42 | Show starting (and ending) at quarter past: everything following York Report | No | [C](https://ury.org.uk/schedule/2010/w42/)
2013 | 10 | Show straddling day (>24 hours) | No | [C](https://ury.org.uk/schedule/2013/w10/)
2017 | 11 | Show straddling day (>24 hours): note 2013-site incorrectly culls 7am and 8am rows here despite Jailbreak being on! | No | [C(B)](https://ury.org.uk/schedule/2017/w11/)
2017 | 00 | Week out of range; should 404 | **Crashes** | [C](https://ury.org.uk/schedule/2010/w00/)
2017 | 54 | Week out of range; should 404 | **Crashes** | [C(S)](https://ury.org.uk/schedule/2010/w54/)
9999 | 52 | Latest schedule permitted by 2016-site routing; crashes 2013-site.  Crashes due to failed URL reverse | [Fixed](https://github.com/UniversityRadioYork/2016-site/pull/64/commits/a2a563153e07e90cff92a53cc0e20cf2b194ebef) | [C(S)](https://ury.org.uk/schedule/1000/w1/)
 

In sitelinks, 'C' links to the current schedule on 2013-site.  Note that 2013-site starts schedules at 7am, while 2016-site starts at 6am.  A (B) next to a site link means the schedule shown is known bugged.  An (S) next to one means the schedule crashes the website.  Eventually I'll upload renders of 2016-site's schedules; these will be tagged with the date.

Known issues coming from this:

- [ ] 404 when week is out of range
- [x] Handle schedule overlaps elegantly
- [x] Grey out prev/next schedule URLs when they can't be reversed (1000, 9999)

MattWindsor91 committed:
Don't emit previous/next URLs if out of range

Fixes schedule crashes on week 1 of year 1000, and week 52 of
year 9999.

MattWindsor91 commented:
@LordAro: what specifically would you want adding to this PRQ to get it viable?  (Presumably fixing the schedule overlap crashes---I'm unsure which is the best way to handle that, whether to just say 'nope I can't render this schedule' or to do surgery on the overlapping shows to make one appear to end before the other).

LordAro commented:
I would say that it should be able to handle overlapping schedules, indeed. Doesn't have to handle it well, just not crash (since currently that results in a blank page which is bad for everyone involved)

MattWindsor91 commented:
Another bug: we don't route `w01` through `w09`, which is a regex bug on my part.

**EDIT**: fixed, but the router can route weeks that don't exist, like `w00` and `w54`, which crash 2016-site and do weird things on 2013-site.

MattWindsor91 committed:
Refactor slice creator; handle overlaps

This simplifies what used to be FillTimeslotSlice, adds truncating
of overlapping shows, and renames it.

Fixes crashes during overlapping shows.

MattWindsor91 committed:
Remove unused weekdayFromVars.

MattWindsor91 committed:
Add route compatibility

Routes .../w01/ through .../w09/, as well as day schedule routes,
now match.

MattWindsor91 commented:
@mstratford, @Brookke: I'm a little confused as to which parts of 2016-site have responsibility over which bits of the schedule generating process.  Is schedule table generation the role of the model or the controller? Where does schedule slice (eg. filled with timeslots) generation go: the same place, or somewhere else?  If they both go into different places, what's the distinction?

Currently on 2016-site, the model doesn't seem to be doing a lot of lifting, and I'm not sure if that's because it's intended to be a thin wrapper around `myradio-go` or if the things that have been implemented just don't need a lot of model code yet.

In my PRQ, some things are lumped into `structs` (which seems like an odd name for a layer in the architecture, because controllers and models have structs too), and some in `controllers`, and the only distinction between the two is that things that depend on `utils` are in `controllers` to break a cyclic dependency, and everything else is in `structs` because I didn't know if it was part of the model or controller or neither.

I don't want this merged until the hierarchy is correct.

MattWindsor91 commented:
(For comparison: 2013-site's main hierarchy was based on area of the website _then_ MVC^, not MVC then area of the website.  I legitimately don't know which of the two is better.)

^ though it calls what 2016-site calls controllers 'views' for some pyramid-based reason

MattWindsor91 commented:
@mstratford, @Brookke: is there an architecture document or design sketch anywhere for 2016-site?

Brookke commented:
@MattWindsor91 There isn't afaik

MattWindsor91 committed:
Codacy 😓

MattWindsor91 committed:
Rename ScheduleWeek controller

Mainly for simplicity, since the same controller will probably serve all of the schedule (for now).

MattWindsor91 committed:
Fix broken StripHtml reference in tests

MattWindsor91 committed:
Move schedule tabulating logic to model

The distribution of work between model and controller is mostly
set up so that the model doesn't need to know about
routing and config structure, but handles everything else.

LordAro commented on views/schedule_week.tmpl:
```diff
@@ -0,0 +1,71 @@
+{{define "content"}}
```
I think we were replacing tab indentation with 2 spaces?

LordAro commented on server.go:
```diff
@@ -40,9 +42,20 @@ func NewServer(c *structs.Config) (*Server, error) {
+	/* NOTE: NewScheduleWeekController assumes the following routes are
+	   installed BEFORE it is called:
+
+	   - 'timeslot' */
```
If this is all it depends on, comment can probably be squished up

MattWindsor91 replied:
Yeah, this is mainly in case other routes start referencing each other

LordAro commented on models/schedule_item.go:
```diff
@@ -0,0 +1,196 @@
+package models
+
+// TODO(CaptainHayashi): this probably doesn't belong in structs.
```
good thing it's not in structs then :)

LordAro reviewed (commented):
LGTM (mostly). I agree that there is some stuff that could probably be in myradio-go (mostly the date stuff in utils, probably). I wouldn't worry too much about the size of the code, it is a rather complex thing and as long as it's well documented (looks like it is to me) i think it'll be fine

MattWindsor91 committed:
Tidy up week schedule, address some issues

Most of this comes from comments @LordAro made -- thanks!  =)

There are a few things here that I thought I'd already committed,
but apparently not.  Maybe I failed at Git last time?

MattWindsor91 commented on structs/config.go:
```diff
@@ -4,15 +4,16 @@ package structs
 type Server struct {
 	Address string `toml:"address"`
 	Port    int    `toml:"port"`
-	Timeout int    `toml:"timout"`
```
Tim!  Out!!

MattWindsor91 commented on structs/config.go:
```diff
@@ -33,10 +34,21 @@ type PageContext struct {
 	Name string `toml:"name"`
-	Url  string `toml:"url"`
+	URL  string `toml:"url"`
 }
```
This and the timeout change might be overstepping it a bit in this PRQ

LordAro commented:
Goodenough.jpg
